R/env.R

```diff
@@ -0,0 +1,702 @@
+#'   to remove.
+#' @return The input object \code{env}, with its associated
+#'   environment modified in place.
+#' @export
+#' @examples
+#' dict <- stats::setNames(letters, letters)
+#' env_bind(environment(), dict)
+#' env_has(environment(), letters)
+#'
+#' # env_unbind() will remove bindings:
+#' env_unbind(environment(), letters)
+#' env_has(environment(), letters)
+#'
+#' # env_eliminate() removes all bindings:
+#' parent <- env_new(env_empty(), list(foo = "a"))
+#' env <- env_new(parent, list(foo = "b"))
+#' env_eliminate(env, "foo")
+#' env_sees(env, "foo")
+env_unbind <- function(env = env_caller(), nms) {
+  env_ <- rlang::env(env)
+  rm(list = nms, envir = env)
+  env
+}
+#' @rdname env_unbind
+#' @export
+env_eliminate <- function(env = env_caller(), nms) {
+  env_ <- rlang::env(env)
+  while(any(env_sees(env_, nms))) {
+    rm(list = nms, envir = env, inherits = TRUE)
+  }
+  env
+}
+
+#' Is an object a dictionary?
+#'
+#' @param x An object to test.
+#' @export
+is_dict <- function(x) {
+  if (!length(x)) {
+    return(!is.null(x))
+  }
+
+  nms <- names(x)
+  if (is.null(nms)) {
+    return(FALSE)
+  }
+
+  is_bad_nm <- is.na(nms) | nms == "" | duplicated(nms)
+  !any(is_bad_nm)
+}
+
+#' Does an environment have or see bindings?
+#'
+#' These functions are vectorised predicates. \code{env_has()} queries
+#' whether an environment owns bindings personally, while
+#' \code{env_sees()} will also return \code{TRUE} for the bindings
+#' that are owned by one of the parent environments.
+#'
+#' @inheritParams env_unbind
+#' @return A logical vector as long as \code{nms}.
+#' @export
+#' @examples
+#' parent <- env_new(env_empty(), list(foo = "foo"))
+#' env <- env_new(parent, list(bar = "bar"))
+#'
+#' # env does not own `foo` but sees it in its parent environment:
+#' env_has(env, "foo")
+#' env_sees(env, "foo")
+env_has <- function(env = env_caller(), nms) {
+  env_ <- rlang::env(env)
+  vapply_lgl(nms, exists, envir = env_, inherits = FALSE)
+}
+#' @rdname env_has
+#' @export
+env_sees <- function(env = env_caller(), nms) {
+  env_ <- rlang::env(env)
+  vapply_lgl(nms, exists, envir = env_, inherits = TRUE)
+}
+
+#' Get an object from a scope.
+#'
+#' \code{env_grab()} looks up an object binded in \code{env} without
+#' looking in its parent. \code{env_fetch()} will return objects from
+#' the parents as well. In that respect, \code{env_grab()} is like
+#' \code{env_has()} and \code{env_fetch()} is like \code{env_sees()}.
+#'
+#' @inheritParams env
+#' @param nm The name of a binding.
+#' @return An object if it exists. Otherwise, throws an error.
+#' @examples
+#' parent <- env_new(env_empty(), list(foo = "foo"))
+#' env <- env_new(parent, list(bar = "bar"))
+#'
+#' # This throws an error because `foo` is not directly defined in env:
+#' # env_grab(env, "foo")
+#'
+#' # However `foo` can be fetched in the parent environment:
+#' env_fetch(env, "foo")
+env_grab <- function(env = env_caller(), nm) {
```

Comment:
Because it would be misleadingly symmetric to `env_set()`. Should it be `env_get()` anyway?

At first `env_set()` was `set_env()` for that reason (setting the env of an object rather than setting something in the env) but it seemed wrong so I changed it.